librbd, tools: migrate from boost::variant to std::variant

tchaikov opened the pull request:
migrate from boost::variant to std::variant

Complete migration started in commit 017f3339c, replacing boost::variant with std::variant throughout the librbd codebase. This change is part of our ongoing effort to reduce third-party dependencies by leveraging C++ standard library alternatives where possible.

Benefits include:
- Improved code readability and maintainability
- Reduced external dependency surface
- More consistent API usage with other components





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

idryomov commented on src/librbd/MirroringWatcher.cc:
```diff
@@ -108,8 +108,8 @@ void MirroringWatcher<I>::handle_notify(uint64_t notify_id, uint64_t handle,
 
-  apply_visitor(watcher::util::HandlePayloadVisitor<MirroringWatcher<I>>(
-                  this, notify_id, handle), notify_message.payload);
+  visit(watcher::util::HandlePayloadVisitor<MirroringWatcher<I>>(
+	  this, notify_id, handle), notify_message.payload);
```
It's better qualify the call -- I hope we get rid of remaining `using namespace std` instances in random header files one day:
```suggestion
  std::visit(watcher::util::HandlePayloadVisitor<MirroringWatcher<I>>(
	       this, notify_id, handle), notify_message.payload);
```
This applies throughout the PR.

tchaikov replied:
i think that this is how ADL works, and we are not relying `using namespace std` in this case.  see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82619

tchaikov replied:
> It's better qualify the call -- I hope we get rid of remaining `using namespace std` instances in random header files one day:
> 

we've removed a lot of `using namespace std` in header files before. see #42742 and #44736. and the only leftover is:
```console
$ git grep 'using namespace std;' | grep '\.h'
src/tools/ceph_dedup/common.h:using namespace std;
```
#63145 was created to address it.

tchaikov replied:
fair enough. i intended to create a minimal change which only addresses the `boost::variant` migration instead of changing whatever not looks nice. but if qualifying the `visit()` with `std::` helps, i am not against including the change in my PR. will do it.

idryomov commented on src/tools/rbd_mirror/ServiceDaemon.cc:
```diff
@@ -184,6 +184,13 @@ void ServiceDaemon<I>::remove_callout(int64_t pool_id, uint64_t callout_id) {
   schedule_update_status();
 }
 
+std::ostream& operator<<(std::ostream& out, const AttributeValue& value) {
```
This is compensating for `std::variant` lacking https://www.boost.org/doc/libs/1_88_0/doc/html/boost/operator_1_3_43_5_4_1_3.html, right?

tchaikov replied:
correct.

idryomov commented on src/tools/cephfs_mirror/ServiceDaemon.cc:
This belongs to CephFS, not sure if it was meant to be included in an RBD PR.

tchaikov replied:
indeed. will drop them from this PR. thank you for pointing this out.

idryomov commented on src/tools/cephfs_mirror/Types.h:
ditto -- this belongs to CephFS

tchaikov commented:
changelog:

- dropped the changes in cephfs
- amended the commit message to explain the unqualified call
- amended the commit message to explain why we are adding operator<< for `AttributeValue`.

tchaikov commented:
changelog:

- replaced the `visit()` calls with `std::visit()` to improve the readability
- removed the "unqualifying" explanation in commit message
- explained why we are using `std::visit()` instead of `visit()` in the migration in the commit message.

idryomov commented:
jenkins test make check arm64

idryomov commented:
Ignoring unrelated `run-tox-mgr-dashboard-lint` failure in `make check` and `make check arm64` -- it's affecting all PRs.

tchaikov commented:
> Ignoring unrelated `run-tox-mgr-dashboard-lint` failure in `make check` and `make check arm64` -- it's affecting all PRs.

thanks. filed https://tracker.ceph.com/issues/71251
